pkg/scheduler/queue/tree_queue.go

```diff
@@ -52,241 +145,133 @@ func (q *TreeQueue) IsEmpty() bool {
-// the receiver/parent node name will create a child node of the same name as the parent.
-func (q *TreeQueue) getOrAddNode(childPath QueuePath) (*TreeQueue, error) {
-	if len(childPath) == 0 {
-		return q, nil
-	}
+func (n *Node) dequeue() (QueuePath, any) {
+	var v any
+	var childPath QueuePath
 
-	if q.childQueueMap == nil {
-		q.childQueueMap = make(map[string]*TreeQueue, 1)
-	}
+	path := QueuePath{n.name}
 
-	var childQueue *TreeQueue
-	var ok bool
-	if childQueue, ok = q.childQueueMap[childPath[0]]; !ok {
-		// no child node matches next path segment
-		// create next child before recurring
-		childQueue = NewTreeQueue(childPath[0])
+	if n.IsEmpty() {
+		return path, nil
+	}
 
-		// add new child queue to ordered list for round-robining;
-		// in order to maintain round-robin order as nodes are created and deleted,
-		// the new child queue should be inserted directly before the current child
-		// queue index, essentially placing the new node at the end of the line
-		if q.currentChildQueueIndex == localQueueIndex {
-			// special case; cannot slice into childQueueOrder with index -1
-			// place at end of slice, which is the last slot before the local queue slot
-			q.childQueueOrder = append(q.childQueueOrder, childQueue.name)
+	var checkedAllNodes bool
+	var dequeueNode *Node
+	// continue until we've found a value or checked all nodes that need checking
+	for v == nil && !checkedAllNodes {
+		dequeueNode, checkedAllNodes = n.dequeueAlgorithm.dequeueGetNode(n)
+		var deletedNode bool
+		// dequeuing from local queue
+		if dequeueNode == n {
+			if n.localQueue.Len() > 0 {
+				// dequeueNode is self, local queue non-empty
+				if elt := n.localQueue.Front(); elt != nil {
+					n.localQueue.Remove(elt)
+					v = elt.Value
+				}
+			}
+		} else if dequeueNode == nil {
+			// no dequeue-able child found; reset checked children to 0,
+			// as we won't update state before moving on
+			n.childrenChecked = 0
+			return path, v
 		} else {
-			// insert into order behind current child queue index
-			q.childQueueOrder = append(
-				q.childQueueOrder[:q.currentChildQueueIndex],
-				append(
-					[]string{childQueue.name},
-					q.childQueueOrder[q.currentChildQueueIndex:]...,
-				)...,
-			)
-			// update current child queue index to its new place in the expanded slice
-			q.currentChildQueueIndex++
+			// dequeue from a child
+			childPath, v = dequeueNode.dequeue()
+			// if the dequeue node is empty _after_ dequeuing, delete it from children
+			if dequeueNode.IsEmpty() {
+				// removing an element sets our position one step forward;
+				// tell state to reset it to original queuePosition
+				delete(n.queueMap, dequeueNode.Name())
+				deletedNode = n.dequeueAlgorithm.deleteChildNode(n, dequeueNode)
+			}
 		}
-
-		// attach new child queue to lookup map
-		q.childQueueMap[childPath[0]] = childQueue
+		n.dequeueAlgorithm.dequeueUpdateState(n, v, deletedNode)
```

Comment:
I think we can ditch `deleteChildNode` altogether.
What's happening here with the current implementation is:
1. we delete the node from the tree
2. we delete the node from the algorithm's node set with `deleteChildNode`
3. we [potentially] update the the algorithm's state based on the `nodeDeleted` bool during `dequeueUpdateState`

Steps 2 & 3 are the same step - it's a bit sketchy / awkward to have operations like "deleting an option from the rotation" and "updating metadata associated with deleting an option from the rotation" be two separate method calls - those are the same operation and if we look at the round robin and TQA implementations for each method I think that's clear.

I am guessing you may have had the idea to make deleting separate from reporting dequeue state since conceivably you could come across an empty node and want to delete it, but you didn't dequeue anything so you aren't ready to break out and update state yet - but we should not be able to get into that state as long as we perform the deletion every time we empty out the last item from a node.

We would change `if dequeueNode.IsEmpty() { // do all the delete stuff }` to 
```golang
if v != nil && dequeueNode.IsEmpty() {
    delete(n.queueMap, dequeueNode.Name())
    // set flag for dequeueUpdateState to use
    // break or even return if you want for clarity, but v not being nil will break us out anyway
}
```

Reply:
I see; to clarify -- you're suggesting, instead of defining a `deleteChildNode`:
1. Set a flag that a node _should_ be deleted
2. Handle all the actual deletion + position updating in `dequeueUpdateState` based on the flag.

?

Reply:
> 1. Set a flag that a node should be deleted
> 2. Handle all the actual deletion + position updating in dequeueUpdateState based on the flag.

I was actually thinking
1. Delete the node from the *tree*  - just the `delete(n.queueMap, dequeueNode.Name())` & set a flag that it _was_ deleted (from the tree, not the state yet)
2. Handle the deletion in the algo state's fields in dequeueUpdateState based on the flag.

But I think either works?
If we went with your way and have moved _everything_ into the algo with `dequeueUpdateState` doing the deletion on both on the node and the state, then the flag could even be a bit more generally named, like `isEmpty`. Fits along the lines of "the tree just reports information, while the algorithm decides what to do with it"

Reply:
> If we went with your way and have moved everything into the algo with dequeueUpdateState doing the deletion on both on the node and the state, then the flag could even be a bit more generally named, like isEmpty. Fits along the lines of "the tree just reports information, while the algorithm decides what to do with it"

I like this! I'll work on making the change.